File: doc-examples/jdbc-example-records/src/main/java/example/Book.java
package example;

import edu.umd.cs.findbugs.annotations.Nullable;
import io.micronaut.data.annotation.DateCreated;
import io.micronaut.data.annotation.GeneratedValue;
import io.micronaut.data.annotation.Id;
import io.micronaut.data.annotation.MappedEntity;

import java.util.Date;

@MappedEntity
record Book(
        @Id @GeneratedValue @Nullable Long id,
        @DateCreated Date dateCreated,
        String title,
        int pages) {
}

File: doc-examples/jdbc-example-records-kotlin/src/main/kotlin/example/Book.kt
package example

import io.micronaut.core.annotation.Nullable
import io.micronaut.data.annotation.*
import java.util.*

@MappedEntity // (1)
data class Book(@Id
                @field:Id @GeneratedValue var id: Long?, // (2)
                @DateCreated @Nullable var dateCreated: Date? = null,
                var title: String,
                var pages: Int = 0) {
    constructor(title: String, pages: Int) : this(null, null, title, pages)
}
